### PR TITLE
Entity Cast Organize

### PR DESCRIPTION
**Pull request by mostafakhudair**

### CODE REVIEW

**Comment by iRedds** [comment]
I don't think it is correct to rename AbstractCast to BaseCast.
You should also make changes to the UserGuide.

**Comment by iRedds** [comment]
Since you have created a separate Entity namespace, it would be logical to move the exception and the Entity class itself there.
But the latter will lead to BC.

**Comment by mostafakhudair** [comment]
I just realized that all abstracted classes named Base...., that's why I renamed it to BaseCast

**Comment by mostafakhudair** [comment]
I'm planing to move Entity class to the new namespace, but yet.

**Comment by MGatner** [comment]
I think this is a better structure. I'm on mobile so can't confirm but... these haven't been released yet, right?

**Comment by mostafakhudair** [comment]
Yes, That's why I did it now to avoid BC later

**Comment by mostafakhudair** [comment]
@MGatner  we need to make sure is will be merged before upcoming release

**Comment by MGatner** [comment]
*sigh* I mean I'm glad we're getting these taken care while we can, but now this is delaying release. Let's do what we can to get this wrapped up ASAP

File: system/Exceptions/CastException.php [hunk 1]
```diff
@@ -43,20 +43,4 @@ public static function forInvalidJsonFormatException(int $error)
 				return new static(lang('Cast.jsonErrorUnknown'));
 		}
 	}
```
Comment (paulbalandan) [inline system/Exceptions/CastException.php:45]: Shouldn't this exception be gone in favor of the other?

**Review by paulbalandan — COMMENTED** [review]
I'd like to get @iRedds' review comments on this as he's the original author of the addition.

**Comment by iRedds** [comment]
@paulbalandan, In principle, I have already expressed my opinion above. 
This PR does not change the functionality that I proposed.

**Comment by paulbalandan** [comment]
> @paulbalandan, In principle, I have already expressed my opinion above.
> This PR does not change the functionality that I proposed.

I am just wondering if you have more to add.

**Comment by MGatner** [comment]
I think we should go ahead and move Entity. You can leave `CodeIgniter\Entity` as a deprecated class extension of `CodeIgniter\Entity\Entity`. I'm open to other ideas though.

**Comment by paulbalandan** [comment]
I would like to see `CodeIgniter\Entity` as an empty file with only `class_exists(...)` and `class_alias(...)` to the new namespace. I just thinking if this would be autoloaded by our `Autoloader` class.

File: system/Exceptions/CastException.php (hunk 1 above)
Comment (mostafakhudair) [reply system/Exceptions/CastException.php:45]: This would be BC, this exception were added before version 4.1

Comment (paulbalandan) [reply system/Exceptions/CastException.php:45]: If I'm not mistaken, system exceptions are strictly for internal use and is not covered by BC. Wait, I'll check.

Comment (paulbalandan) [reply system/Exceptions/CastException.php:45]: OK. That was for language strings. Do add `@deprecated` tag for the class instead.

**Comment by mostafakhudair** [comment]
What if we cleaned current Entity class from all methods and extends it from CodeIgniter\Entity\Entity, and also mark at as deprecated class?

**Comment by MGatner** [comment]
Can we get this squashed? Also happy to squash it on merge, but we don't need 30 commits of the same thing :O

**Comment by mostafakhudair** [comment]
sorry for that, but i cannot push from my local to this PR any more, but once test is fixed, i'll create new clean PR with one commit :D

**Pull request closed without merging**